### ISSUE TITLE
Add focus ring to all components for improved accessibility

### ISSUE THREAD

**Issue by juanruitina** [ubuntu/yaru.dart#979]
To be compliant with WCAG Level AA, focused elements should be clearly different from their unfocused state. We seem to have inherited some insufficient focus states from Material. This was already flagged up for buttons in https://github.com/ubuntu/yaru.dart/issues/849. I suggest a general approach to the focus indicators in Yaru components: using solid 2px-wide focus rings.

I'm fully aware implementing this for all components would be a big piece of work and take time, but it would be great if we could agree on the approach.

## Rationale

[Success Criterion 2.4.13 Focus Appearance](https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance.html) provides some guidance on good focus indicators:

> When the keyboard [focus indicator](https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance.html#dfn-focus-indicator) is visible, an area of the focus indicator meets all the following:
> 
> - is at least as large as the area of a 2 [CSS pixel](https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance.html#dfn-css-pixel) thick [perimeter](https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance.html#dfn-perimeter) of the unfocused component or sub-component, and
> - has a contrast ratio of at least 3:1 between the same pixels in the focused and unfocused states.

Changing background on focus, as most components in Yaru.dart do, is generally insufficient to achieve the 3:1 contrast requirement. On top of that, we rely on background changes for hover.

The Yaru GTK theme does use focus rings. These rings use transparency (code [here](https://github.com/ubuntu/yaru/blob/7db6f52042fa3c5bb0bb7f843848c51569e21e94/gtk/upstream/gtk-4.0/Default/_colors.scss#L74-L103) and [here](https://github.com/ubuntu/yaru/blob/7db6f52042fa3c5bb0bb7f843848c51569e21e94/gtk/src/default/gtk-3.0/_tweaks.scss#L161-L204)): this makes the rings lighter visually and perhaps more appealing for some users, but also result in insufficient contrast when high contrast mode is off, <3:1 (rings are slightly more opaque when high contrast mode is on).

I suggest that we don't make such differentiation with high contrast mode: a 3:1 contrast ratio will arguably benefit all users, and the focus ring appears only after a specific keyboard event, which was one of the concerns brought up in https://github.com/ubuntu/yaru/issues/1769. It will also make these focus rings unquestionably compliant with WCAG, as the orange accent color has a contrast >3:1 on white and light gray.

## Specification

Here my proposal on how to address focus rings systematically:

* Focus ring should be the theme’s accent colour (orange by default).
* Standalone components (e.g. regular buttons) and any component with a solid background should have a 2px outside outline, with a 2px offset.
* Components that don’t have a solid background and are within bordered containers (e.g. menu items) should have a 2px inside outline.
* In form fields, outline should replace existing borders (2px, inside).
* Outline should match component shape (for example, border radius).

## Mockup

<img width="1314" alt="Mockup with different components, with and without focus ring" src="https://github.com/user-attachments/assets/576f9a08-d411-4d70-877a-15a7aa9cff56" />

## Potential issues

Given orange is our default accent colour, there are some corner cases in which the focus ring may dilute the meaning of semantic red, as flagged in https://github.com/ubuntu/yaru.dart/issues/843. A possible mitigation would be using an error icon.

**Comment by Feichtmeier** [comment]
HI @juanruitina thanks for your very detailed issue!

We had a discussion about this with one of your previous collegues (elio), and @Jupi007 and me came to the insight, that this is not possible with the flutter material buttons. (which we use and apply a theme, unlike the checkboxes and switches, which @Jupi007  completely rewrote from scratch).
Which does not mean that this is not possible in general, but we then need to create completely new buttons, or make an upstream contribution at flutter. This is the upstream issue: https://github.com/flutter/flutter/issues/120425
. 

Also since I guess if buttons have a focus ring, the users expects them on every widget. 
Personally I don't think I can currently create such a large contribution but maybe @Jupi007 can or one of canonicals engineers?
A small part of your requests are thicker focus rings for textfields,   this is easily possible with the theme, we can do this easily

**Comment by juanruitina** [comment]
Thank you for providing context, @Feichtmeier: I'll discuss it with the team.

**Comment by juanruitina** [comment]
I updated slightly my proposal incorporating feedback from my colleague @lukasserelis (also heavy user of keyboard navigation!). We agree focus rings should always have the same colour, regardless of the component's, so it's perceivable as user jumps from one element to the next. Even if it might result in weird combinations in some corner cases (e.g. orange-red in destructive actions), functionality trumps aesthetics in this case as this is not visible by default. This will also make implementation simpler.

**Comment by ashuntu** [comment]
@Feichtmeier would you be open to a solution using https://github.com/ubuntu/yaru.dart/pull/1042?

I could wrap most Yaru widgets with `YaruFocusBorder`, which would cover the vast majority of focus-able widgets that the App Center and Installer would use. I think the only real area that Yaru doesn't cover is regular, non-split buttons, but I could also add those. Then, any areas that Yaru doesn't cover can be covered by just using `YaruFocusBorder` as needed in projects themselves.
